ospfd/ospf_ext.c

```diff
@@ -0,0 +1,1777 @@
+	struct ext_itf *exti;
+	uint32_t label;
+
+	/* Process Neighbor only when its state is NSM Full */
+	if (nbr->state != NSM_Full)
+		return;
+
+	/* Get interface information for Segment Routing */
+	if ((exti = lookup_ext_by_ifp(oi->ifp)) == NULL) {
+		zlog_warn(
+			"EXT (ospf_ext_link_nsm_change) Cannot get Extended "
+			"information from OI(%s)",
+			IF_NAME(oi));
+		return;
+	}
+
+	if (oi->area == NULL || oi->area->ospf == NULL) {
+		zlog_warn(
+			"EXT (ospf_ext_link_nsm_change) Cannot refer to "
+			"OSPF from OI(%s)",
+			IF_NAME(oi));
+		return;
+	}
+
+	/* Keep Area information in combination with SR info. */
+	exti->area = oi->area;
+	OspfEXT.area = oi->area;
+
+	/* Process only Adjacency/LAN SID */
+	if (exti->stype == PREF_SID)
+		return;
+
+	switch (oi->state) {
+	case ISM_PointToPoint:
+		/* Segment ID is an Adjacency one */
+		exti->stype = ADJ_SID;
+
+		/* Set Extended Link TLV with link_id == Nbr Router ID */
+		set_ext_link(exti, OSPF_IFTYPE_POINTOPOINT, nbr->router_id,
+			     oi->address->u.prefix4);
+
+		/* Set Extended Link Adjacency SubTLVs, backup first */
+		label = get_ext_link_label_value();
+		set_adj_sid(exti, true, label, SID_LABEL);
+		label = get_ext_link_label_value();
+		set_adj_sid(exti, false, label, SID_LABEL);
+		/* And Remote Interface address */
+		set_rmt_itf_addr(exti, nbr->address.u.prefix4);
+
+		break;
+
+	case ISM_DR:
+		/* Segment ID is a LAN Adjacency for the DR only */
+		exti->stype = LAN_ADJ_SID;
+
+		/* Set Extended Link TLV with link_id == DR */
+		set_ext_link(exti, OSPF_IFTYPE_BROADCAST, DR(oi),
+			     oi->address->u.prefix4);
+
+		/* Set Extended Link Adjacency SubTLVs, backup first */
+		label = get_ext_link_label_value();
+		set_lan_adj_sid(exti, true, label, SID_LABEL, nbr->router_id);
+		label = get_ext_link_label_value();
+		set_lan_adj_sid(exti, false, label, SID_LABEL, nbr->router_id);
+
+		break;
+
+	case ISM_DROther:
+	case ISM_Backup:
+		/* Segment ID is an Adjacency if not the DR */
+		exti->stype = ADJ_SID;
+
+		/* Set Extended Link TLV with link_id == DR */
+		set_ext_link(exti, OSPF_IFTYPE_BROADCAST, DR(oi),
+			     oi->address->u.prefix4);
+
+		/* Set Extended Link Adjacency SubTLVs, backup first */
+		label = get_ext_link_label_value();
+		set_adj_sid(exti, true, label, SID_LABEL);
+		label = get_ext_link_label_value();
+		set_adj_sid(exti, false, label, SID_LABEL);
+
+		break;
+
+	default:
+		if (CHECK_FLAG(exti->flags, EXT_LPFLG_LSA_ENGAGED)) {
+			ospf_ext_link_lsa_schedule(exti, FLUSH_THIS_LSA);
+			UNSET_FLAG(exti->flags, EXT_LPFLG_LSA_ENGAGED);
+			UNSET_FLAG(exti->flags, EXT_LPFLG_LSA_ACTIVE);
+		}
+		return;
+	}
+
+	/* flood this links params if everything is ok */
+	SET_FLAG(exti->flags, EXT_LPFLG_LSA_ACTIVE);
+	if (OspfEXT.enabled) {
+		if (CHECK_FLAG(exti->flags, EXT_LPFLG_LSA_ENGAGED))
+			ospf_ext_link_lsa_schedule(exti, REFRESH_THIS_LSA);
+		else
+			ospf_ext_link_lsa_schedule(exti, REORIGINATE_THIS_LSA);
+	}
+
+	return;
```

Comment:
I don't see the point of all these blank return statements at the end of void functions.